snuba/pipeline/composite.py

```diff
@@ -0,0 +1,274 @@
+        )
+
+
+class CompositeDataSourcePlan(NamedTuple):
+    """
+    Intermediate data structure maintained when visiting and
+    transforming the query data source.
+
+    We cannot use a Composite plan itself as an intermediate
+    structure thus the need of this class.
+    This is because the data source type is a bit less
+    restrictive in this class and we would have to instantiate
+    an execution strategy that is not used.
+    """
+
+    translated_source: Union[ClickhouseQuery, CompositeQuery[Table], JoinClause[Table]]
+    root_processors: Optional[SubqueryProcessors] = None
+    aliased_processors: Optional[Mapping[str, SubqueryProcessors]] = None
+
+    @classmethod
+    def from_simple_query_plan(
+        cls, query_plan: ClickhouseQueryPlan
+    ) -> CompositeDataSourcePlan:
+        return CompositeDataSourcePlan(
+            translated_source=query_plan.query,
+            root_processors=SubqueryProcessors(
+                plan_processors=query_plan.plan_query_processors,
+                db_processors=query_plan.db_query_processors,
+            ),
+        )
+
+    @classmethod
+    def from_composite_subquery(
+        cls, query_plan: CompositeQueryPlan
+    ) -> CompositeDataSourcePlan:
+        return CompositeDataSourcePlan(
+            translated_source=query_plan.query,
+            root_processors=query_plan.root_processors,
+            aliased_processors=query_plan.aliased_processors,
+        )
+
+
+class CompositeDataSourcePlanner(DataSourceVisitor[CompositeDataSourcePlan, Entity]):
+    """
+    Plans the DataSource (the from clause) of a Composite query by
+    visiting each node as the data source can be composite itself.
+
+    At each step an intermediate data structure containing the
+    translated query and the query processors to run in the execution
+    strategy is produced.
+    These intermediate structures are merged together and produce the
+    CompositeQueryPlan.
+    """
+
+    def __init__(self, request_settings: RequestSettings) -> None:
+        self.__settings = request_settings
+
+    def _visit_simple_source(self, data_source: Entity) -> CompositeDataSourcePlan:
+        # We are never supposed to get here.
+        raise ValueError("Cannot translate a simple Entity")
+
+    def _visit_join(self, data_source: JoinClause[Entity]) -> CompositeDataSourcePlan:
+        join_visitor = JoinDataSourcePlanner(self.__settings)
+        plan = data_source.accept(join_visitor)
+
+        assert isinstance(plan.translated_source, JoinClause)
+        return CompositeDataSourcePlan(
+            translated_source=plan.translated_source,
+            aliased_processors=plan.processors,
+        )
+
+    def _visit_simple_query(
+        self, data_source: ProcessableQuery[Entity]
+    ) -> CompositeDataSourcePlan:
+        return CompositeDataSourcePlan.from_simple_query_plan(
+            _plan_simple_query(data_source, self.__settings)
+        )
+
+    def _visit_composite_query(
```

Comment:
How would we end up here - wouldn't we be looking at the data source inside the composite query at this point? Or if there's another scenario to support could we add a test for it?

Reply:
This is meant to be for this query:
```
SELECT avg(releases) FROM (
   SELECT e.project_id, uniq(e.release) FROM 
      sentry_local e INNER JOIN groupedmessage_local g ON e.group_id = g.id
   GROUP BY e.prject_id
)
```
The data source in the composite query is a composite query itself.

I did not add a specific test for this because of the massive verbosity of the AST of that query, since the nested query was tested and the join was tested. 
I can add one if you prefer.

Reply:
I see - how many levels of nesting will be supported in this way? Just the one, or are multiple levels of nesting to be supported too?

Reply:
as many as the client wants. For as long as a join query is not further nested (join only allow simple subqueries, and this is by design otherwise we would not be able to express them as predetermined relationships on a graph. the structure of a query does not have anything other than simple queries as nodes of a join).

Though if somebody creates a query with more than 2 nesting levels, I think they would be doing something wrong.